### ISSUE TITLE
Poedit crashes when pressing F10, Space

### ISSUE THREAD

**Issue by rillig** [vslavik/poedit#322]
OS: Windows 10
Poedit: Version 2.0beta2 (4723)

To reproduce:
1.  Start Poedit without any file
2.  Press the F10 key. This will select the File menu
3.  Press the Space key. This will crash Poedit

These are the simplest steps to reproduce. My actual workflow looked like this:

1. Start Poedit with an existing translation
2. Focus the translation text field
3. Press AltGr+8, which on a German keyboard layout usually inserts an opening square bracket
4. Press Space

After step 3, the main menu is focused, which is weird. Also, no character is inserted into the translation text field. This means I cannot type any characters requiring the AltGr key.

**Comment by vslavik** [comment]
Thanks! Looks like these are two unrelated bugs :(

**Comment by vslavik** [comment]
> After step 3, the main menu is focused, which is weird. Also, no character is inserted into the translation text field. This means I cannot type any characters requiring the AltGr key.

I’m not seeing the behavior you describe (with German keyboard). Yes, the menu is highlighted due to a mCtrl bug, but the text is also typed as it should. Are you sure you had focus in the text entry area at the time?

**Comment by vslavik** [comment]
Could you please try this replacement executable to see if it fixes the issue (it does for me, but see above, there seem to be *some* discrepancies between our situations)? Or if there aren’t any other keyboard/menu related bugs?

https://jumpshare.com/v/QRtupakRD9Gy3CgRun8f

Thanks!

**Comment by rillig** [comment]
> Are you sure you had focus in the text entry area at the time?

Veeery good question. I am not sure, and therefore I tried again. This time, the `[` was inserted, as you expected. I had probably clicked into the English field instead of the German one.

With the replacement executable, the crash doesn't happen anymore.  The menu is still selected.

Thank you for the quick fix, and for the whole program after all. :)

**Comment by vslavik** [comment]
> The menu is still selected.
>
> Thank you for the quick fix

That crash is a superficial triviality, it’s the bad keyboard input I’m concerned about — and what you’re saying is that it *wasn’t* fixed at all.

At the risk of repeating myself, are you *sure*? Because doing what you claim to do here, namely using AltGr+8 with the German keyboard (this matters, US keyboard behaves differently!), *does not* select the menu in this build.

If you’re still seeing unintended menu invocation, can you please describe what *exactly* you do, in the that test build, to trigger it? Thanks!

**Comment by rillig** [comment]
The exact steps are:

1.  Start Windows. I'm on version 10.0.14393.
2.  Start Poedit from the start menu. I'm on version 2.0beta2 (4723), as instructed by your above comment.
3.  Hold down the <kbd>AltGr</kbd> key. The menu letters get underlined.
4.  Release the <kbd>AltGr</kbd> key. The file menu gets a light blue background.

It's as simple as that.

For comparison, I tried Audacity, since it also uses wxWidgets. But that one doesn't show the effect.
Switching the keyboard to English also shows the effect.
Switching to a newly created local Windows account shows the effect.
Even if I turn off all other Windows applications, the effect is still there.

**Comment by vslavik** [comment]
> Hold down the <kbd>AltGr</kbd> key. The menu letters get underlined.
4.  Release the <kbd>AltGr</kbd> key. The file menu gets a light blue background.

I thought you were saying that the menu still gets wrongly selected *while typing [ using AltGr* (as was the case before the test build above). That seems to be very different thing from the above. Am I missing something?

I'll retest tomorrow but I would swear I checked that just pressing different Alt keys (i.e. LeftAlt vs RightAlt vs AltGr), without doing anything else, behaves differently as it should.

> since it also uses wxWidgets. 

wx has no relevance to this.

**Comment by rillig** [comment]
Sorry, please give me a last try on this:

1.  Start Poedit.
1.  Hold down the <kbd>AltGr</kbd> key. The menu letters get underlined.
2.  Press <kbd>8</kbd>. Nothing changes visually.
3.  Release <kbd>8</kbd>. Nothing changes visually.
4.  Release the <kbd>AltGr</kbd> key. The file menu gets a light blue background.

**Comment by vslavik** [comment]
(Apologies in advance for stupid questions that will follow and will feel like I treat you like a moron. That’s absolutely not it — but *something* is very, *very* wrong here, there are vast unexplainable differences in basic behavior between what I see in the same version of Windows, with the same keyboard, and between what you report. I’m trying to verify if my fix is correct, so I need to get to the bottom of this. So again, my apologies. Just trying to rule out things out.)

(1) First of, please use this build: https://jumpshare.com/v/D1Y7dS0IgUEm5tzyeU5e — it’s identical to the previous one, but adds a “TEST BUILD” string to the titlebar *after you open a file*, so that you can be sure you’re using the correct version. This is to ensure that you’re not still accidentally using the old build; it certainly would be one of the two simplest explanations.

(2) Per your latest step-by-step instructions, you’re trying to use AltGr+8 *when there’s nothing to type in*: you say to just "start Poedit”, which shows the welcome screen. There’s no place to type, so how can you test typing `[`? Please open a PO file and put focus into editing area for testing (as that’s where AltGr behavior matters). This is a "just in case” thing, because I see no wrong behavior on the welcome screen either here.

(3) This isn’t how Poedit or any native Windows app (try e.g. Notepad) behave: 
> Hold down the AltGr key. The menu letters get underlined.

The accelerators aren’t underlined until after you *release* the Alt key. Do you use any custom settings somewhere to change that?

(4) Do we understand each other about the difference between the Alt key (on the left side) and AltGr (on the right side *if you use German keyboard*)? Alt is supposed to behave like this, AltGr is not (i.e. the bug is that on Windows 10, Poedit/mCtrl treat AltGr equally to Alt).

(5) Please verify that you have in fact the German keyboard (not just German-language keyboard, but German *layout* keyboard) set when testing this, because English keyboards don’t have AltGr. This the 2nd simplest explanation for the discrepancy, that you accidentally test with a wrong keyboard.

(6) Please try some other key combinations too, such as AltGr+E (for €) or AltGr+Q (for @).

(7) Please try in Notepad too, just in case.

If all these checks pass, then I am at complete loss :( It just *can’t* be this radically different...

**Comment by vslavik** [comment]
Also, you *are* able type using AltGr now, correct? There were three issues: crash (you reported fixed), selection of menu when it should (you reported still there) and inability to type e.g. “[“ with AltGr (you didn’t explicitly say either way).

**Comment by rillig** [comment]
> (Apologies in advance for stupid questions that will follow and will feel like I treat you like a moron.

I absolutely don't feel treated in any bad way. Narrowing down the issue with further precise questions is the only thing we can do to narrow down the problem.

> (1) First of, please use this build: https://jumpshare.com/v/D1Y7dS0IgUEm5tzyeU5e — it’s identical to the previous one, but adds a “TEST BUILD” string to the titlebar after you open a file, so that you can be sure you’re using the correct version. This is to ensure that you’re not still accidentally using the old build; it certainly would be one of the two simplest explanations.

Succeeded. The "** TEST BUILD" appears.

> (2) Per your latest step-by-step instructions, you’re trying to use AltGr+8 when there’s nothing to type in: you say to just "start Poedit”, which shows the welcome screen. There’s no place to type, so how can you test typing [?

With these step-by-step instructions I wanted to demonstrate that the menu gets the focus when it shouldn't. Since it doesn't matter whether I press the <kbd>AltGr</kbd> key in a text field or anywhere else, I thought having as few steps as possible would be best for reproducing the issue.

So this was not about being able to *type* something but just to reproduce the main menu focus. Since that is already unusual I thought fixing this issue would also fix the typing issue. That's why I concentrated on this one.

> Please open a PO file and put focus into editing area for testing (as that’s where AltGr behavior matters). This is a "just in case” thing, because I see no wrong behavior on the welcome screen either here.

> (3) This isn’t how Poedit or any native Windows app (try e.g. Notepad) behave:
> 
> Hold down the AltGr key. The menu letters get underlined.
> The accelerators aren’t underlined until after you release the Alt key. Do you use any custom settings somewhere to change that?

I know that this is not the normal behavior. I tried Notepad and Audacity and Chrome and Poedit 1.8.11, and all these behave as expected. Only Poedit 2.0 doesn't, and I have no idea why.

> (4) Do we understand each other about the difference between the Alt key (on the left side) and AltGr (on the right side if you use German keyboard)? Alt is supposed to behave like this, AltGr is not (i.e. the bug is that on Windows 10, Poedit/mCtrl treat AltGr equally to Alt).

I know that LeftAlt is different from AltGr.
I know that AltGr is indistinguishable from Alt+Strg on the WM_KEYDOWN layer.
I don't know what you mean by "the bug is that on Windows 10, Poedit/mCtrl treat AltGr equally to Alt". Is this "bug" a known one or the one we're currently hunting?

> (5) Please verify that you have in fact the German keyboard (not just German-language keyboard, but German layout keyboard) set when testing this, because English keyboards don’t have AltGr. This the 2nd simplest explanation for the discrepancy, that you accidentally test with a wrong keyboard.

I have a Logitech keyboard with German labels on all the keys.
The keyboard layout in Windows is German/Germany.
As I already said above, when I switch the Windows keyboard layout to English (US), pressing <kbd>AltGr</kbd> also focuses the main menu.

> (6) Please try some other key combinations too, such as AltGr+E (for €) or AltGr+Q (for @).

All the AltGr key combinations have the same effect: I can type, but as soon as I release the <kbd>AltGr</kbd> key, the main menu gets the focus.

> (7) Please try in Notepad too, just in case.

I did. Notepad behaves normally, as does Poedit 1.8.11.

> If all these checks pass, then I am at complete loss :( It just can’t be this radically different...

Can you think of anything that changed between 1.8.11 and 2.0beta2?

> Also, you are able type using AltGr now, correct? There were three issues:
> crash (you reported fixed),

No crash anymore.

> selection of menu when it should (you reported still there)

When it *shouldn't*. Yes, this is still there in 2.0beta2, 2.0beta2-nocrash and 2.0beta2-testbuild. But not in 1.8.11.

> and inability to type e.g. “[“ with AltGr (you didn’t explicitly say either way).

I *am* able to type. It's just that when I release <kbd>AltGr</kbd>, that the main menu gets the focus. When I press <kbd>AltGr</kbd> again, the focus is back in the text field and I can continue typing.

To rule out any other processes, I terminated all processes except Explorer.exe and Poedit.exe, so that should not be an issue. Yesterday I installed Visual Studio, so I will try to find out which WM_KEY messages are sent around. I still have no idea *why* this behaves differently.

A big thank you for all the support.

**Comment by vslavik** [comment]
> (3) This isn’t how Poedit or any native Windows app (try e.g. Notepad) behave:
> 
> Hold down the AltGr key. The menu letters get underlined.
> The accelerators aren’t underlined until after you release the Alt key. Do you use any custom settings > somewhere to change that?
> I know that this is not the normal behavior. I tried Notepad and Audacity and Chrome and Poedit 1.8.11, and all these behave as expected. Only Poedit 2.0 doesn't, and I have no idea why.

What does it *mean* “as expected” *exactly*? This issue very much requires exact, precise formulations, it keeps getting more and more confusing otherwise.

Are you saying that the menu items are *not* underlined in Notepad when you *hold* Alt (*not* AltGr!) (and are only underlined after you *release* it), but they *are* in Poedit 2 as soon as you press and still *hold* Alt (*not* AltGr!)?

Because that’s absolutely not what I’m seeing at all. In fact, part of the confusion was my use of Parallels — where holding Alt (*not* AltGr) doesn’t underline accels, whereas a physical USB keyboard does (here too). So I eliminated one discrepancy… except that your statement above, directly responding to me talking about exactly this issue, seems to contradict this as well as your previous statements. That makes debugging extremely hard :(

> Can you think of anything that changed between 1.8.11 and 2.0beta2?

I am fully aware of what caused it and need help with understanding the cause of the bug. I *already fixed the bug* as far as I can tell — while I could reproduce the reported bugs without issues locally, I no longer can with the fix and Poedit behaves 100% identically to other native applications. 

To be extra sure, I borrowed a Windows USB keyboard and plugged it in directly, eliminating any possibility of the input being messed up by Parallels. Still behaves *identically* in Poedit and everywhere else (with the difference in when menu items are highlighted during pressing Alt between hardware and Parallels keyboards).

**What I need to do is understand why the fix *which is working here* mysteriously does *NOT* work for *you*.** 

About *that* I’m at lost. The principal difference between Alt and AltGr is that [Alt sends WM_SYSKEYDOWN while AltGr sends WM_KEYDOWN](https://blogs.msdn.microsoft.com/murrays/2015/04/20/hot-keys-and-altgr/) — they can be distinguished easily enough. And it’s not like Windows would send radically different messages on different machines with the exact same OS version. So *why* does the fix not work in your case?

I’m so confused it’s not even funny.

Here’s one more test to try: Is there any difference between how LeftCtrl+LeftAlt+E behaves in Notepad and in the Poedit testbuild? AltGr is actually implemented as LeftCtrl+Alt in Windows and using the above combination types € in the same way as AltGr+E does — they are identical at the low level keyboard handling.

**Comment by rillig** [comment]
>> (3) This isn’t how Poedit or any native Windows app (try e.g. Notepad) behave:
>> 
>> Hold down the AltGr key. The menu letters get underlined.
>> The accelerators aren’t underlined until after you release the Alt key. Do you use any custom settings >> somewhere to change that?
> I know that this is not the normal behavior. I tried Notepad and Audacity and Chrome and Poedit 1.8.11, and all these behave as expected. Only Poedit 2.0 doesn't, and I have no idea why.
> What does it mean “as expected” exactly? This issue very much requires exact, precise formulations, it keeps getting more and more confusing otherwise.

With "as expected" I meant that these applications do not focus the main menu when I press <kbd>AltGr</kbd>. Since pretty much all applications act this way, this is what I expected.

> Are you saying that the menu items are not underlined in Notepad when you hold Alt (not AltGr!) (and are only underlined after you release it), but they are in Poedit 2 as soon as you press and still hold Alt (not  AltGr!)?

No, I'm not saying that. I never said anything about the <kbd>Alt</kbd> key. I only said something about <kbd>AltGr</kbd>.

But now that you mention it, I tried out what happens in Notepad, Poedit 1.8.11 and Poedit 2.0beta2:

Notepad and Poedit 1.8.11:

| Action | Menu letters | File menu<br>background color | Text cursor |
|--- | --- | --- | --- |
| 0. Start | not underlined | white | blinks |
| 1. KeyDown LeftAlt | underlined | white | blinks |
| 2. KeyUp LeftAlt | underlined | blue | does not blink |
| 3. KeyDown LeftAlt | not underlined | white | blinks |
| 4. KeyRepeat LeftAlt | not underlined | white | blinks |
| 5. KeyUp LeftAlt | not underlined | white | blinks |

Poedit 2.0beta2 behaves the same, unless you wait 3 seconds between step 3 KeyDown LeftAlt and 4 KeyUp LeftAlt. Then the following happens:

| Action | Menu letters | File menu<br>background color | Text cursor | Remarks
|--- | --- | --- | --- | --- |
| 0. Start | not underlined | white | blinks | 
| 1. KeyDown LeftAlt | underlined | white | blinks | 
| 2. KeyUp LeftAlt | underlined | blue | does not blink | 
| 3. KeyDown LeftAlt | not underlined | white | blinks | 
| 4. KeyRepeat LeftAlt | underlined | white | blinks | different
| 5. KeyUp LeftAlt | underlined | blue | does not blink | different

Maybe this is a clue to find the actual cause.

> Because that’s absolutely not what I’m seeing at all. In fact, part of the confusion was my use of Parallels — where holding Alt (not AltGr) doesn’t underline accels, whereas a physical USB keyboard does (here too). So I eliminated one discrepancy… except that your statement above, directly responding to me talking about exactly this issue, seems to contradict this as well as your previous statements. That makes debugging extremely hard :(

> Here’s one more test to try: Is there any difference between how LeftCtrl+LeftAlt+E behaves in Notepad and in the Poedit testbuild? AltGr is actually implemented as LeftCtrl+Alt in Windows and using the above combination types € in the same way as AltGr+E does — they are identical at the low level keyboard handling.

*   In Notepad, LeftCtrl+LeftAlt+E types €, without underlining the main menu letters.
*   In Notepad, LeftAlt+LeftCtrl+E types €, while temporarily underlining the main menu letters.
    After releasing all keys, the focus is still in the text field.
*   Poedit 1.8.11 behaves the same as Notepad.
*   In Poedit 2.0beta2, LeftCtrl+LeftAlt already underlines the main menu letters, which it shouldn't.
    After releasing all keys, the main menu has the focus.
*   In Poedit 2.0beta2, LeftCtrl+LeftAlt+E types € and underlines the main menu letters.
    After releasing all keys, the main menu has the focus.

I hope these cases clear up the situation. If you give me some pointers where to look in the code, I can do that. I have successfully checked out Poedit master, built it and it's running. My build shows the same weird behavior as the testbuild version you sent me.

**Comment by vslavik** [comment]
> Maybe this is a clue to find the actual cause.

Nah, that’s *another* subtle bug in the [mCtrl menubar](http://www.mctrl.org) control that you discovered (thanks!).

I’m *terribly* sorry, I just discovered what the cause of the discrepancy is: my mind-boggling colossal stupidity. It would be impressive if it weren’t sad that I wasted so much of your time like this — and I can’t even blame it on the flu I was down with most of the week because I wasn’t nearly enough sick to be delirious.

You see, I sent you a replacement `Poedit.exe`. But the fix, being for mCtrl, was in `mctrl.dll`, which I never sent you… so of course you didn’t see any changes.

😐🔫

Here’s the DLL that presumably fixes it for you once you replace Poedit’s standard one with it: https://jumpshare.com/v/S26mnp6WRzXLVRoOwGj7

Sorry again for such an inexcusable waste of your time.
